src/binder_sms.c

```diff
@@ -285,29 +297,42 @@ binder_sms_sca_query_cb(
     gpointer user_data)
 {
     BinderSmsCbData* cbd = user_data;
+    BinderSms* self = cbd->self;
     ofono_sms_sca_query_cb_t cb = cbd->cb.sca_query;
     struct ofono_error err;
 
     if (status == RADIO_TX_STATUS_OK) {
-        if (resp == RADIO_RESP_GET_SMSC_ADDRESS) {
+        guint32 code = self->interface_aidl == RADIO_MESSAGING_INTERFACE ?
+            RADIO_MESSAGING_RESP_GET_SMSC_ADDRESS :
+            RADIO_RESP_GET_SMSC_ADDRESS;
+        if (resp == code) {
             if (error == RADIO_ERROR_NONE) {
                 GBinderReader reader;
                 const char* smsc;
 
                 gbinder_reader_copy(&reader, args);
-                smsc = gbinder_reader_read_hidl_string_c(&reader);
+                if (self->interface_aidl == RADIO_AIDL_INTERFACE_NONE) {
+                    smsc = gbinder_reader_read_hidl_string_c(&reader);
+                } else {
+                    smsc = gbinder_reader_read_string16(&reader);
+                }
                 if (smsc) {
                     struct ofono_phone_number sca;
+                    const char* str = smsc;
 
-                    if (smsc[0] == '+') {
-                        smsc++;
+                    if (str[0] == '+') {
+                        str++;
                         sca.type = OFONO_NUMBER_TYPE_INTERNATIONAL;
                     } else {
                         sca.type = OFONO_NUMBER_TYPE_UNKNOWN;
                     }
-                    g_strlcpy(sca.number, smsc, sizeof(sca.number));
+                    g_strlcpy(sca.number, str, sizeof(sca.number));
                     DBG("csca_query_cb: %s, %d", sca.number, sca.type);
                     cb(binder_error_ok(&err), &sca, cbd->data);
+
+                    if (self->interface_aidl == RADIO_MESSAGING_INTERFACE) {
+                        g_free((char*)smsc);
```

Comment:
Free of a `const char*` with a cast? No.

Reply:
Fixed this and couple of other similar cases in other files.